### PR TITLE
Feature/5945 rat profile

### PR DESCRIPTION
**Pull request by KaiTeuber**
## Description
Show Antigentest-Profil with entered user data and QR Code

## Link to Jira
https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-6620

## Screenshots
Simulator Screen Shot - iPhone 12 Pro - 2021-04-27 at 14 27 14
Simulator Screen Shot - iPhone 12 Pro - 2021-04-27 at 14 27 45
Simulator Screen Shot - iPhone 12 Pro - 2021-04-27 at 14 27 49
Simulator Screen Shot - iPhone 12 Pro - 2021-04-27 at 14 27 11

## Notice
At the Moment vCard Version 3.0 is used, changes to Version 4.0 will be done in an other ticket

### CODE REVIEW

File: src/xcode/ENA/ENA/Source/Views/Common/CloseBarButtonItem.swift [hunk 1]
```diff
@@ -32,11 +24,37 @@ class CloseBarButtonItem: UIBarButtonItem {
...
 	func didTap() {
 		onTap()
 	}
 
-	let onTap: () -> Void
+	// MARK: - Private
+
+	private let onTap: () -> Void
+
+	private func setupNormalMode(_ mode: Mode) {
```
Comment (pascal-brause) [inline src/xcode/ENA/ENA/Source/Views/Common/CloseBarButtonItem.swift:41]:
```suggestion
	private func setup(_ mode: Mode) {
```

File: src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/CreateAntigenTestProfileViewModel.swift [hunk 2]
```diff
@@ -15,7 +15,14 @@ final class CreateAntigenTestProfileViewModel {
 		self.antigenTestProfile = AntigenTestProfile()
 
 		// this is only for coordinator testing, remove later
```
Comment (pascal-brause) [inline src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/CreateAntigenTestProfileViewModel.swift:17]: should that be removed or not?

File: src/xcode/ENA/ENA/Source/Scenes/AntigenTestProfile/Cells/QRCodeCellViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+			postalAddress.street = antigenTestProfile.addressLine ?? ""
+			postalAddress.postalCode = antigenTestProfile.zipCode ?? ""
+			contact.postalAddresses = [CNLabeledValue(label: CNLabelHome, value: postalAddress)]
+		}
+
+		do {
+			let vCardData = try CNContactVCardSerialization.data(with: [contact])
+			return vCardData
+		} catch {
+			Log.error("Failed to create vCard data with antigenTestProfile input")
```
Comment (pascal-brause) [inline src/xcode/ENA/ENA/Source/Scenes/AntigenTestProfile/Cells/QRCodeCellViewModel.swift:72]:
```suggestion
			Log.error("Failed to create vCard data with antigenTestProfile input with error: \(error)")
```

File: src/xcode/ENA/ENA/Source/Scenes/AntigenTestProfile/AntigenTestProfileViewController.swift [hunk 4]
```diff
@@ -62,20 +67,145 @@ class AntigenTestProfileViewController: UIViewController, FooterViewHandling, Di
...
+				tableView.trailingAnchor.constraint(equalTo: view.trailingAnchor)
+
+			]
+		)
+
+		// observer tableView scrolling to move the background in sync
+		tableContentObserver = tableView.observe(\UITableView.contentOffset, options: .new) { [weak self] _, change in
+			guard let self = self,
+				  let yOffset = change.newValue?.y else {
+				return
```
Comment (pascal-brause) [inline src/xcode/ENA/ENA/Source/Scenes/AntigenTestProfile/AntigenTestProfileViewController.swift:196]: perhaps leave a Log entry here?

File: src/xcode/ENA/ENA/Source/Scenes/AntigenTestProfile/AntigenTestProfileViewController.swift [hunk 5]
```diff
@@ -62,20 +67,145 @@ class AntigenTestProfileViewController: UIViewController, FooterViewHandling, Di
...
+	func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+		viewModel.numberOfItems(in: AntigenTestProfileViewModel.TableViewSections.map(section))
+	}
+
+	// MARK: - UITableViewdelegate
+
+	func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
+		guard didCalculateGradientHeight == false,
+			  AntigenTestProfileViewModel.TableViewSections.map(indexPath.section)  == .QRCode else {
+			return
```
Comment (pascal-brause) [inline src/xcode/ENA/ENA/Source/Scenes/AntigenTestProfile/AntigenTestProfileViewController.swift:85]: Perhaps leave a log entry here?

File: src/xcode/ENA/ENA/Source/Extensions/UIImage+QRCode.swift [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+	class func qrCode(
+		with string: String,
+		encoding: String.Encoding = .shiftJIS,
+		size: CGSize = CGSize(width: 400, height: 400),
+		qrCodeErrorCorrectionLevel: MappedErrorCorrectionType = .medium
+	) -> UIImage? {
+		/// Create data from string which will be feed into the CoreImage Filter
+		guard let data = string.data(using: encoding) else {
+			return nil
```
Comment (pascal-brause) [inline src/xcode/ENA/ENA/Source/Extensions/UIImage+QRCode.swift:17]: Perhaps leave a log entry here?

File: src/xcode/ENA/ENA/Source/Extensions/UIImage+QRCode.swift [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+		/// Create CoreImage Filter to create QR-Code
+		guard let filter = CIFilter(name: "CIQRCodeGenerator") else {
+			return nil
+		}
+		filter.setValue(data, forKey: "inputMessage") /// Feed data into Filter
+		filter.setValue(qrCodeErrorCorrectionLevel.mappedValue, forKey: "inputCorrectionLevel") /// Set ErrorCorrectionLevel
+
+		guard let image = filter.outputImage else {
+			return nil
```
Comment (pascal-brause) [inline src/xcode/ENA/ENA/Source/Extensions/UIImage+QRCode.swift:28]: Perhaps leave a log entry here?

File: src/xcode/ENA/ENA/Source/Extensions/UIImage+QRCode.swift [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+		qrCodeErrorCorrectionLevel: MappedErrorCorrectionType = .medium
+	) -> UIImage? {
+		/// Create data from string which will be feed into the CoreImage Filter
+		guard let data = string.data(using: encoding) else {
+			return nil
+		}
+
+		/// Create CoreImage Filter to create QR-Code
+		guard let filter = CIFilter(name: "CIQRCodeGenerator") else {
+			return nil
```
Comment (pascal-brause) [inline src/xcode/ENA/ENA/Source/Extensions/UIImage+QRCode.swift:22]: Perhaps leave a log entry here?

File: src/xcode/ENA/ENA/Source/Scenes/AntigenTestProfile/AntigenTestProfileViewModel.swift [hunk 9]
```diff
@@ -3,31 +3,159 @@
...
+	func numberOfItems(in section: TableViewSections) -> Int {
+		switch section {
+		default:
+			return 1
+		}
+	}
+
 	// MARK: - Private
 
+	enum TableViewSections: Int, CaseIterable {
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/AntigenTestProfile/AntigenTestProfileViewModel.swift:107]:
```suggestion
	enum TableViewSection: Int, CaseIterable {
```

File: src/xcode/ENA/ENA/Source/Scenes/AntigenTestProfile/Cells/QRCodeCellViewModel.swift [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+			  )
+		else {
+			Log.error("Failed to create QRCode image for vCard data")
+			return UIImage()
+		}
+		return QRCodeImage
+	}
+
+	// create vCard data, CNContactVCardSerialization creates a version 3.0 output
+	var vCardData: Data {
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Scenes/AntigenTestProfile/Cells/QRCodeCellViewModel.swift:44]: Is this still needed. Because there is a v4 implementation?

**Review by ArturFriesen — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/Scenes/AntigenTestProfile/Cells/QRCodeCellViewModel.swift (hunk 10 above)
Comment (KaiTeuber) [reply src/xcode/ENA/ENA/Source/Scenes/AntigenTestProfile/Cells/QRCodeCellViewModel.swift:44]: yes in this branch - will get removed by the V4 PR

File: src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/CreateAntigenTestProfileViewModel.swift (hunk 2 above)
Comment (KaiTeuber) [reply src/xcode/ENA/ENA/Source/Scenes/CreateAntigenTestProfile/CreateAntigenTestProfileViewModel.swift:17]: At the moment it is required, but should be removed within the create profile branch

File: src/xcode/ENA/ENA/Source/Scenes/AntigenTestProfile/AntigenTestProfileViewController.swift [hunk 11]
```diff
@@ -62,20 +67,145 @@ class AntigenTestProfileViewController: UIViewController, FooterViewHandling, Di
...
+		}
+
+		let cellRect = tableView.rectForRow(at: indexPath)
+		let result = view.convert(cellRect, from: tableView)
+		backgroundView.gradientHeightConstraint.constant = result.midY
```
[the remark this answers is not in the document]
Comment (KaiTeuber) [reply src/xcode/ENA/ENA/Source/Scenes/AntigenTestProfile/AntigenTestProfileViewController.swift:90]: yes, that's why we recalculate things in `tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell` if a `traitCollectionDidChange` got called

File: src/xcode/ENA/ENA/Source/Scenes/AntigenTestProfile/AntigenTestProfileViewModel.swift [hunk 12]
```diff
@@ -3,31 +3,159 @@
...
+		default:
+			return 1
+		}
+	}
+
 	// MARK: - Private
 
+	enum TableViewSection: Int, CaseIterable {
+		case header
+		case QRCode
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Scenes/AntigenTestProfile/AntigenTestProfileViewModel.swift:109]: Just a small thingy, can we keep camel case here? qrCode

File: src/xcode/ENA/ENA/Source/Scenes/AntigenTestProfile/AntigenTestProfileViewController.swift (hunk 4 above)
Comment (KaiTeuber) [reply src/xcode/ENA/ENA/Source/Scenes/AntigenTestProfile/AntigenTestProfileViewController.swift:196]: I added a comment - a log it's helpful here

File: src/xcode/ENA/ENA/Source/Scenes/AntigenTestProfile/AntigenTestProfileViewController.swift (hunk 5 above)
Comment (KaiTeuber) [reply src/xcode/ENA/ENA/Source/Scenes/AntigenTestProfile/AntigenTestProfileViewController.swift:85]: I added a comment - a log it's helpful here

File: src/xcode/ENA/ENA/Source/Scenes/AntigenTestProfile/AntigenTestProfileViewModel.swift [hunk 13]
```diff
@@ -3,31 +3,159 @@
...
+			format: AppStrings.ExposureSubmission.AntigenTest.Profile.dateOfBirthFormatText,
+			DateFormatter.localizedString(from: dateOfBirth, dateStyle: .medium, timeStyle: .none)
+		)
+	}
+
+	private var phoneNumber: String? {
+		return antigenTestProfile.phoneNumber
+	}
+
+	private var emailAdress: String? {
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Scenes/AntigenTestProfile/AntigenTestProfileViewModel.swift:149]: emailAddress

File: src/xcode/ENA/ENA/Source/Scenes/AntigenTestProfile/AntigenTestProfileViewController.swift (hunk 4 above)
Comment (KaiTeuber) [reply src/xcode/ENA/ENA/Source/Scenes/AntigenTestProfile/AntigenTestProfileViewController.swift:196]: I improved the comment above, a log might not help here

File: src/xcode/ENA/ENA/Source/Scenes/AntigenTestProfile/Cells/QRCodeCellViewModel.swift [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+	// MARK: - Init
+
+	init(
+		antigenTestProfile: AntigenTestProfile,
+		backgroundColor: UIColor,
+		boarderColor: UIColor
+	) {
+		self.antigenTestProfile = antigenTestProfile
+		self.backgroundColor = backgroundColor
+		self.boarderColor = boarderColor
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Scenes/AntigenTestProfile/Cells/QRCodeCellViewModel.swift:19]: borderColor

File: src/xcode/ENA/ENA/Source/Scenes/AntigenTestProfile/Cells/QRCodeCellViewModel.swift (hunk 3 above)
Comment (KaiTeuber) [reply src/xcode/ENA/ENA/Source/Scenes/AntigenTestProfile/Cells/QRCodeCellViewModel.swift:72]: At the moment we do not have an error here - I could look for it.
But the vCard creation will change and that code gets removed next.

File: src/xcode/ENA/ENA/Source/Scenes/AntigenTestProfile/Cells/SimpleTextCellViewModel.swift [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+
+	init(
+		backgroundColor: UIColor?,
+		textColor: UIColor? = nil,
+		textAlignment: NSTextAlignment? = nil,
+		text: String? = nil,
+		attributedText: NSAttributedString? = nil,
+		topSpace: CGFloat,
+		font: UIFont,
+		boarderColor: UIColor = .clear,
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Scenes/AntigenTestProfile/Cells/SimpleTextCellViewModel.swift:20]: borderColor

**Review by pascal-brause — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/Extensions/UIImage+QRCode.swift (hunk 6 above)
Comment (KaiTeuber) [reply src/xcode/ENA/ENA/Source/Extensions/UIImage+QRCode.swift:17]: done

**Review by naveeddotio — APPROVED** [review]

**Pull request merged**